pkg/p2p/message_router.go

```diff
@@ -0,0 +1,182 @@
+// limitations under the License.
+
+package p2p
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/ticdc/pkg/security"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// MessageRouter is used to maintain clients to all the peers in the cluster
+// that the local node needs to communicate with.
+type MessageRouter interface {
+	// AddPeer should be invoked when a new peer is discovered.
+	AddPeer(id NodeID, addr string)
+	// RemovePeer should be invoked when a peer is determined to
+	// be permanently unavailable.
+	RemovePeer(id NodeID)
+	// GetClient returns a MessageClient for `target`. It returns
+	// nil if the target peer does not exist. The returned client
+	// is canceled if RemovePeer is called on `target`.
+	GetClient(target NodeID) *MessageClient
+	// Close cancels all clients maintained internally.
+	Close()
+	// Wait waits for all clients to exit.
+	Wait()
+	// Err returns a channel to receive errors from.
+	Err() <-chan error
+}
+
+type messageRouterImpl struct {
+	mu         sync.RWMutex
+	addressMap map[NodeID]string
+	clients    map[NodeID]clientWrapper
+
+	wg       sync.WaitGroup
+	isClosed atomic.Bool
+	errCh    chan error
+
+	// read only field
+	credentials  *security.Credential
+	selfID       NodeID
+	clientConfig *MessageClientConfig
+}
+
+// NewMessageRouter creates a new MessageRouter
+func NewMessageRouter(selfID NodeID, credentials *security.Credential, clientConfig *MessageClientConfig) MessageRouter {
+	return &messageRouterImpl{
+		addressMap:   make(map[NodeID]string),
+		clients:      make(map[NodeID]clientWrapper),
+		errCh:        make(chan error, 1), // one error at most
+		credentials:  credentials,
+		selfID:       selfID,
+		clientConfig: clientConfig,
+	}
+}
+
+type clientWrapper struct {
+	*MessageClient
+	cancelFn context.CancelFunc
+}
+
+// AddPeer implements MessageRouter.
+func (m *messageRouterImpl) AddPeer(id NodeID, addr string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.addressMap[id] = addr
+}
+
+// RemovePeer implements MessageRouter.
+func (m *messageRouterImpl) RemovePeer(id NodeID) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	delete(m.addressMap, id)
+	if clientWrapper, ok := m.clients[id]; ok {
+		clientWrapper.cancelFn()
+	}
+}
+
+// GetClient implements MessageRouter. The client will be created lazily.
+func (m *messageRouterImpl) GetClient(target NodeID) *MessageClient {
+	m.mu.RLock()
+	// fast path
+	if cliWrapper, ok := m.clients[target]; ok {
+		m.mu.RUnlock()
+		return cliWrapper.MessageClient
+	}
+
+	// There is no ready-to-use client for target
+	m.mu.RUnlock()
+	// escalate the lock
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// repeats the logic in fast path after escalating the lock, since
+	// the lock was briefly released.
+	if cliWrapper, ok := m.clients[target]; ok {
+		return cliWrapper.MessageClient
+	}
+
+	addr, ok := m.addressMap[target]
+	if !ok {
+		log.Info("failed to create client, no peer",
```

Comment:
Is `log.Warn` better, If `GetClient` returns nil, will the caller retry it later?

Reply:
Yes, the caller will retry if the target is expected to be added.